File: packages/sdk-go/codec/encoder.go
package codec

import (
	"strings"

	"github.com/uts-dot/sdk-go"
	"github.com/uts-dot/sdk-go/types"
)

type Encoder struct {
	buf []byte
}

func NewEncoder() *Encoder {
	return &Encoder{
		buf: make([]byte, 0, 1024),
	}
}

func NewEncoderWithSize(size int) *Encoder {
	return &Encoder{
		buf: make([]byte, 0, size),
	}
}

func (e *Encoder) Bytes() []byte {
	return e.buf
}

func (e *Encoder) ensureCapacity(n int) {
	if cap(e.buf)-len(e.buf) < n {
		newCap := cap(e.buf) * 2
		if newCap < len(e.buf)+n {
			newCap = len(e.buf) + n
		}
		newBuf := make([]byte, len(e.buf), newCap)
		copy(newBuf, e.buf)
		e.buf = newBuf
	}
}

func (e *Encoder) WriteByte(b byte) *Encoder {
	e.ensureCapacity(1)
	e.buf = append(e.buf, b)
	return e
}

func (e *Encoder) WriteBytes(data []byte) *Encoder {
	e.ensureCapacity(len(data))
	e.buf = append(e.buf, data...)
	return e
}

func (e *Encoder) WriteU32(n uint32) *Encoder {
	e.buf = append(e.buf, EncodeU32(n)...)
	return e
}

func (e *Encoder) WriteU64(n uint64) *Encoder {
	e.buf = append(e.buf, EncodeU64(n)...)
	return e
}

func (e *Encoder) WriteLengthPrefixed(data []byte) *Encoder {
	e.WriteU32(uint32(len(data)))
	e.WriteBytes(data)
	return e
}

func (e *Encoder) WriteOp(op types.Op) *Encoder {
	return e.WriteByte(byte(op))
}

func (e *Encoder) WriteVersionedMagic(version byte) *Encoder {
	e.WriteBytes(MagicBytes)
	e.WriteByte(version)
	return e
}

func (e *Encoder) WriteHeader(header *types.DigestHeader) *Encoder {
	e.WriteOp(types.Op(header.Kind))
	e.WriteBytes(header.DigestBytes())
	return e
}

func (e *Encoder) WriteExecutionStep(step types.Step) *Encoder {
	e.WriteOp(step.Op())
	switch s := step.(type) {
	case *types.AppendStep:
		e.WriteLengthPrefixed(s.Data)
	case *types.PrependStep:
		e.WriteLengthPrefixed(s.Data)
	}
	return e
}

func (e *Encoder) WriteForkStep(step *types.ForkStep) error {
	if len(step.Branches) < 2 {
		return uts.NewEncodeError(uts.ErrCodeInvalidData, "FORK step must have at least 2 branches", nil)
	}
	for _, branch := range step.Branches[:len(step.Branches)-1] {
		e.WriteOp(types.OpFORK)
		if err := e.WriteTimestamp(branch); err != nil {
			return err
		}
	}
	return e.WriteTimestamp(step.Branches[len(step.Branches)-1])
}

func (e *Encoder) WritePendingAttestation(att *types.PendingAttestation) error {
	uri := strings.TrimSuffix(att.URI, "/")
	if len(uri) > types.MaxURILen {
		return uts.NewEncodeError(uts.ErrCodeUriTooLong, "URI exceeds maximum length", nil)
	}
	if !types.ValidateURI(uri) {
		return uts.NewEncodeError(uts.ErrCodeInvalidUriChar, "invalid character in URI", nil)
	}
	e.WriteLengthPrefixed([]byte(uri))
	return nil
}

func (e *Encoder) WriteBitcoinAttestation(att *types.BitcoinAttestation) *Encoder {
	e.WriteU32(att.Height)
	return e
}

func (e *Encoder) WriteEASAttestation(att *types.EASAttestation) *Encoder {
	e.WriteU64(att.ChainID)
	e.WriteBytes(att.UID[:])
	return e
}

func (e *Encoder) WriteEASTimestamped(att *types.EASTimestamped) *Encoder {
	e.WriteU64(att.ChainID)
	return e
}

func (e *Encoder) WriteUnknownAttestation(att *types.UnknownAttestation) *Encoder {
	e.WriteBytes(att.Tag[:])
	e.WriteLengthPrefixed(att.Data)
	return e
}

func (e *Encoder) WriteAttestationStep(step *types.AttestationStep) error {
	e.WriteOp(types.OpAttestation)
	inner := NewEncoder()

	switch att := step.Attestation.(type) {
	case *types.PendingAttestation:
		e.WriteBytes(types.PendingTag[:])
		if err := inner.WritePendingAttestation(att); err != nil {
			return err
		}
		e.WriteLengthPrefixed(inner.Bytes())
	case *types.BitcoinAttestation:
		e.WriteBytes(types.BitcoinTag[:])
		inner.WriteBitcoinAttestation(att)
		e.WriteLengthPrefixed(inner.Bytes())
	case *types.EASAttestation:
		e.WriteBytes(types.EASAttestTag[:])
		inner.WriteEASAttestation(att)
		e.WriteLengthPrefixed(inner.Bytes())
	case *types.EASTimestamped:
		e.WriteBytes(types.EASTimestampTag[:])
		inner.WriteEASTimestamped(att)
		e.WriteLengthPrefixed(inner.Bytes())
	default:
		tag := step.Attestation.Tag()
		e.WriteBytes(tag[:])
		e.WriteLengthPrefixed([]byte{})
	}
	return nil
}

func (e *Encoder) WriteStep(step types.Step) error {
	switch s := step.(type) {
	case *types.ForkStep:
		return e.WriteForkStep(s)
	case *types.AttestationStep:
		return e.WriteAttestationStep(s)
	default:
		e.WriteExecutionStep(s)
		return nil
	}
}

func (e *Encoder) WriteTimestamp(ts types.Timestamp) error {
	for _, step := range ts {
		if err := e.WriteStep(step); err != nil {
			return err
		}
	}
	return nil
}

func EncodeDetachedTimestamp(ots *types.DetachedTimestamp) ([]byte, error) {
	enc := NewEncoder()
	enc.WriteVersionedMagic(0x01)
	enc.WriteHeader(ots.Header)
	if err := enc.WriteTimestamp(ots.Timestamp); err != nil {
		return nil, err
	}
	return enc.Bytes(), nil
}

File: packages/sdk-py/src/uts_sdk/_codec/encoder.py
# packages/sdk-py/src/uts_sdk/_codec/encoder.py
"""Binary encoder for OpenTimestamps format."""

from __future__ import annotations

import re
from typing import Self

from uts_sdk._types import (
    AppendStep,
    AttestationStep,
    BitcoinAttestation,
    EASTimestamped,
    EASAttestation,
    ExecutionStep,
    ForkStep,
    PendingAttestation,
    PrependStep,
    UnknownAttestation,
)
from uts_sdk.errors import EncodeError, ErrorCode

from .constants import (
    BITCOIN_TAG,
    EAS_ATTEST_TAG,
    EAS_TIMESTAMPED_TAG,
    MAGIC_BYTES,
    MAX_URI_LEN,
    PENDING_TAG,
)

_SAFE_URL_RE = re.compile(r"^https?://[\s\w./:-]+$")


class Encoder:
    """Binary encoder for OpenTimestamps proof format.

    Example:
        encoder = Encoder()
        encoder.write_u32(42).write_bytes(b"hello")
        data = encoder.to_bytes()
    """

    __slots__ = ("_buffer", "_offset")

    def __init__(self, initial_size: int = 1024) -> None:
        self._buffer = bytearray(initial_size)
        self._offset = 0

    def _ensure_capacity(self, required: int) -> None:
        if self._offset + required > len(self._buffer):
            new_size = max(len(self._buffer) * 2, self._offset + required)
            self._buffer.extend(b"\x00" * (new_size - len(self._buffer)))

    def to_bytes(self) -> bytes:
        return bytes(self._buffer[: self._offset])

    def write_byte(self, value: int) -> Self:
        if not 0 <= value <= 255:
            raise EncodeError(
                ErrorCode.OVERFLOW,
                f"Byte value must be 0-255, got {value}",
            )
        self._ensure_capacity(1)
        self._buffer[self._offset] = value
        self._offset += 1
        return self

    def write_bytes(self, data: bytes) -> Self:
        self._ensure_capacity(len(data))
        self._buffer[self._offset : self._offset + len(data)] = data
        self._offset += len(data)
        return self

    def write_u32(self, value: int) -> Self:
        if value < 0:
            raise EncodeError(
                ErrorCode.NEGATIVE_LEB128_INPUT,
                f"LEB128 only supports non-negative integers, got {value}",
            )
        if value > 0xFFFFFFFF:
            raise EncodeError(
                ErrorCode.OVERFLOW,
                f"Value exceeds maximum for u32: {value}",
            )

        remaining = value
        while True:
            byte = remaining & 0x7F
            remaining >>= 7
            if remaining != 0:
                byte |= 0x80
            self.write_byte(byte)
            if remaining == 0:
                break
        return self

    def write_u64(self, value: int) -> Self:
        if value < 0:
            raise EncodeError(
                ErrorCode.NEGATIVE_LEB128_INPUT,
                f"LEB128 only supports non-negative integers, got {value}",
            )

        remaining = value
        while True:
            byte = remaining & 0x7F
            remaining >>= 7
            if remaining != 0:
                byte |= 0x80
            self.write_byte(byte)
            if remaining == 0:
                break
        return self

    def write_length_prefixed(self, data: bytes) -> Self:
        self.write_u32(len(data))
        self.write_bytes(data)
        return self

    def write_magic(self, version: int = 0x01) -> Self:
        self.write_bytes(MAGIC_BYTES)
        self.write_byte(version)
        return self

    def write_header(self, header) -> Self:
        self.write_byte(header.kind.to_op_code().value)
        self.write_bytes(header.digest)
        return self

    def write_execution_step(self, step: ExecutionStep) -> Self:
        self.write_byte(step.op.value)
        if isinstance(step, (AppendStep, PrependStep)):
            self.write_length_prefixed(step.data)
        return self

    def write_fork_step(self, step: ForkStep) -> Self:
        for branch in step.steps[:-1]:
            self.write_byte(0xFF)
            self.write_timestamp(branch)
        self.write_timestamp(step.steps[-1])
        return self

    def write_pending_attestation(self, att: PendingAttestation) -> Self:
        url = att.url.rstrip("/")
        if len(url) > MAX_URI_LEN:
            raise EncodeError(
                ErrorCode.INVALID_URI,
                f"URL exceeds maximum length of {MAX_URI_LEN}: {len(url)}",
            )
        if not _SAFE_URL_RE.match(url):
            raise EncodeError(
                ErrorCode.INVALID_URI,
                f"Invalid URL format: {url}",
            )
        self.write_length_prefixed(url.encode("utf-8"))
        return self

    def write_bitcoin_attestation(self, att: BitcoinAttestation) -> Self:
        self.write_u32(att.height)
        return self

    def write_eas_attestation(self, att: EASAttestation | EASTimestamped) -> Self:
        self.write_u64(att.chain_id)
        if isinstance(att, EASAttestation):
            self.write_bytes(att.uid)
        return self

    def write_attestation_step(self, step: AttestationStep) -> Self:
        self.write_byte(0x00)

        att = step.attestation
        if isinstance(att, PendingAttestation):
            self.write_bytes(PENDING_TAG)
            inner = Encoder()
            inner.write_pending_attestation(att)
            self.write_length_prefixed(inner.to_bytes())
        elif isinstance(att, BitcoinAttestation):
            self.write_bytes(BITCOIN_TAG)
            inner = Encoder()
            inner.write_bitcoin_attestation(att)
            self.write_length_prefixed(inner.to_bytes())
        elif isinstance(att, EASAttestation):
            self.write_bytes(EAS_ATTEST_TAG)
            inner = Encoder()
            inner.write_eas_attestation(att)
            self.write_length_prefixed(inner.to_bytes())
        elif isinstance(att, EASTimestamped):
            self.write_bytes(EAS_TIMESTAMPED_TAG)
            inner = Encoder()
            inner.write_eas_attestation(att)
            self.write_length_prefixed(inner.to_bytes())
        elif isinstance(att, UnknownAttestation):
            self.write_bytes(att.tag)
            self.write_length_prefixed(att.data)
        else:
            raise EncodeError(
                ErrorCode.GENERAL_ERROR,
                f"Unsupported attestation type: {type(att).__name__}",
            )
        return self

    def write_step(self, step) -> Self:
        if isinstance(step, ForkStep):
            return self.write_fork_step(step)
        elif isinstance(step, AttestationStep):
            return self.write_attestation_step(step)
        else:
            return self.write_execution_step(step)

    def write_timestamp(self, timestamp) -> Self:
        for step in timestamp:
            self.write_step(step)
        return self

    @classmethod
    def encode_detached(cls, ots) -> bytes:
        encoder = cls()
        encoder.write_magic()
        encoder.write_header(ots.header)
        encoder.write_timestamp(ots.timestamp)
        return encoder.to_bytes()
